Fix #5780: Fix develop CI and improve test file check robustness

BenHenning opened the pull request:
## Explanation
Fixes #5780

#5780 was ultimately caused by two things:
1. An issue in how the code coverage utility propagates certain errors for incremental runs.
2. Specific repository issues that can cause failures.

This PR addresses (2) which mitigates (1) in its entirety (for this particular class of issues, that is, incorrectly placed and/or named tests). (1) is to be addressed in #5812.

In order to ensure this problem was fully addressed, `TestFileCheck` has been completely reworked to have a much stricter verification of production and test files. Specifically, it:
- Now knows the production & test directory layouts for each top-level layer package.
- Checks _both_ production and test files to ensure they have corresponding matches.
- Introduces an exemption for tests which should exist but not match their production file names (such as extra tests). We may want to consolidate these in the future, but it's not pressing.
- Checks that exemptions actually match real files--a bunch of the existing exemptions point to non-existent files.
- Presents "could be" suggestions when it encounters a test or production file that might have a match which is in the wrong directory.

Some specific notes:
- The new matching approach has led to a _lot_ of files that have been moved or otherwise fixed:
  - A bunch of admin control, dev options, and help tests weren't in as specific of directories as their production files (or were in the wrong higher-level package).
  - `FaqListActivityTest` didn't match the name of its production file and thus needed to be renamed.
  - `ImageRegionSelectionInteractionViewTest`'s new location is modularized, so it needed a new test declaration. Note that this was largely based on `StateFragmentTest`'s test declaration, so it likely contains redundant dependencies (I didn't check them thoroughly). There's tooling coming in the modularization work to clean these up, so I think it's probably fine.
  - #5806 set the wrong packages for some of the new orchestrator classes (due to the Bazel plugin not knowing where these files are supposed to go), and this was unfortunately missed (yet still built). This has been corrected and the `TestFileCheck` changes will act as a guard against these situations in the future.
  - Some tests had an extra 'rule' in them: `DragDropSortInputHasElementXAtPositionYRuleClassifierProviderTest` and `DragDropSortInputHasElementXBeforeElementYRuleClassifierProviderTest`.
  - `SurveyEventsLoggerTest` and its test target have been moved to the correct `oppialogger` subpackage (which is new).
  - A number of `app` and `domain` tests were moved to the utility layer:
    - `FakeAccessibilityServiceTest`
    - `StringExtensionsTest`
    - Most (all?) of the HTML parsing tests: `CocneptCardTagHandlerTest`, `CustomHtmlContentHandlerTest`, `HtmlParserTest`, `ImageTagHandlerTest`, `ListItemLeadingMarginSpanTest`, `LiTagHandlerTest`, `MathTagHandlerTest`, and `PolicyPageTagHandlerTest`.
    - `PlatformParameterSingletonTest`.
  - `HtmlParserTest` and `ListItemLeadingMarginSpanTest` both depended on bespoke test activities, and trying to make `utility` tests depend on these was challenging. Instead, these have been moved over to using `TestActivity` (for which there's some precedent in non-`app` tests) and the bespoke activities have been removed. This helps with the effort of #3830.
- Instrumentation tests currently need to be marked as 'extra' tests. Fixing this properly will probably require addressing #3436 (that is, having proper mechanisms for detecting permanent exemptions though the mechanism may be different than that issue in particular).
- There are some Java-specific bits built in that can be removed when #1639 is resolved.
- #2532 will simplify some of the extra test directory management. We may want to take this further in the future and remove all top-level layer directories in favor of just UI, service, and scripts. At that point we could also move away from the Maven directory structure, but more thought is needed on that.
- TODO exemptions needed to be updated to wiki page adjustments.
- `UserAppHistoryViewModel` was a file that was introduced a long time ago, but has always been empty. I've been meaning to remove it for a while, but now we finally had a good reason: the new script requires `package` statements in Java and Kotlin files in order to verify that the file was in its correct location (relative to its test/production file), and since this file is empty it has no `package`. It's now been removed.
- `TestFileCheck`'s tests have been completely rewritten. Note that they now focus on testing the utility in isolation rather than using the real exemption script because the new check of verifying that the files exist present a prohibitively difficult restriction on testing `main` directly (i.e. by requiring all these files exist).
- The valid exemption cases for missing tests have been clarified in the Static Analysis Checks wiki page.
- `RecentlyPlayedActivityTest` has been exempted as it was missed originally (which was what ultimately caused #5780) since it is currently incompatible with code coverage.

I recommend looking at the test changes for `TestFileCheckTest` to see the new check output in different failure cases since it better demonstrates what failures this check can now find and report.

This is being verified by both running all tests (which simulates being merged into `develop` and an incremental run after the complete run finishes), specifically:
- Full check: https://github.com/oppia/oppia-android/pull/5814/checks?sha=040baf1e0c72368a03d8123aaf0b4c346e5f297a.
- Incremental check: https://github.com/oppia/oppia-android/pull/5814/checks?sha=67d5f905ed0e67ff4428fc8948c7236fe234d6b7.

## Essential Checklist
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
N/A -- Infrastructure and refactor changes only.

BenHenning committed:
Fix invalid packages.

This also removes an empty file that's unsued: UserAppHistoryViewModel.

These are preempting failures that will be detected by an updated
TestFileCheck script which will be checked in later on this branch.

BenHenning committed:
Move a bunch of test files to matching locations.

This also fixes a few package imports from the previous commit that were
missed.

This addresses all cases where there are prod files without exact
matching test files (outside of exemptions). There are still test files
that aren't matched against prod files to address.

BenHenning committed:
Fix remaining test file issues.

The remaining failures will be exempted.

BenHenning committed:
Revise TestFileCheck.

The new check performs much more thorough analysis to ensure that
exemptions point to valid files, and that directory structures match. It
also ensures that there are pairings for both tests to prod files and
vice versa.

The tests for this check have been completely rewritten, and a new
exemption type was added for extra test suites (such as local tests).

The exemptions proto had a lot of updates needed to remove invalid
references and to exempt these extra test suites.

BenHenning committed:
Lint & static check fixes.

BenHenning committed:
Fix imports in test.

BenHenning committed:
Empty commit to re-trigger CI.

BenHenning commented on scripts/src/java/org/oppia/android/scripts/testfile/TestFileCheck.kt:
```diff
@@ -9,7 +9,7 @@ import java.io.FileInputStream
  *
  * Usage:
- *   bazel run //scripts:test_file_check -- <path_to_directory_root>
+ *   bazel run //scripts:test_file_check -- <path_to_directory_root> <path_to_exemptions>
```
This is wrong--I never ended up changing the arguments.

BenHenning replied:
Fixed in latest.

BenHenning reviewed (commented):
Self-reviewed changes.

BenHenning committed:
Remove incorrect KDoc change.

Addresses self-review comment.

BenHenning commented:
The full run was a success as it didn't run into the same failure as develop. There's one test that requires an exemption which will be added in a subsequent commit.

BenHenning committed:
Exempt failing test for coverage.

BenHenning committed:
Empty commit to re-trigger CI.

BenHenning committed:
Empty commit to re-trigger CI (again).

GitHub seemed to ignore the previous commit.

BenHenning reviewed (commented):
Self-reviewed latest. I also manually checked locally that `RecentlyPlayedActivity` is properly exempted now, as well.

BenHenning commented:
I expect this to pass now, so I think it can be reviewed.

PTAL @adhiamboperes or delegate as needed. This is pretty important as it should stop develop CI from failing.

seanlip reviewed (approved):
LGTM for owners (thanks @BenHenning)!

adhiamboperes commented on wiki/Static-Analysis-Checks.md:
```diff
@@ -138,8 +141,10 @@ exempted_file_path: "app/src/main/java/org/oppia/android/app/activity/ActivityCo
+1. Interface files.
+2. Files with only constants defined (no logic).
+3. Cases when tests fail to pass code coverage.
+4. Cases when a test is verifying additional functionality beyond what can be tested in its usual test suite (for these cases use the 'is_extra_test' property in the exemption structure).
```
It might be worth adding an example here as well. It took me a while to understand why the math parser tests were exempted.

BenHenning replied:
Good point--added an example and made some other small clean-ups in this file.

adhiamboperes reviewed (changes requested):
Thanks @BenHenning! This LGTM, just had one comment.

I am happy to see these improvements - especially catching exemptions that don't correspond to actual files, and improved specificity of the failure messages.

BenHenning committed:
Update Static-Analysis-Checks.md

Adds example + cleans up some of the other code blocks (for better alignment and syntax highlighting).

BenHenning reviewed (commented):
Thanks @seanlip & @adhiamboperes!

I addressed your comment @adhiamboperes, PTAL.

Also, self-reviewed the latest changes.

BenHenning commented:
Odd CI flake--restarting everything.

Also, I'm enabling auto-merge since I think the main comment to verify should be non-controversial, but I'm happy to make additional changes as needed. :)

BenHenning commented:
Ah, unfortunate. TODOs need to be regenerated for the updated wiki page.

adhiamboperes reviewed (approved):
LGTM, thanks!

BenHenning committed:
Correct TODO exemptions in updated wiki page.

BenHenning commented:
Thanks @adhiamboperes! Just waiting on a passing CI now.

BenHenning reviewed (commented):
Self-reviewed TODO exemptions update due to wiki page changes.

BenHenning commented:
GitHub is having a LOT of spurious failures today. Most of these are coming from trying to download `Buf` from a GitHub release URL.
